### DISCUSSION TITLE
Multiple Choice Knapsack Problem

### DISCUSSION THREAD

**Discussion by Eddie-Hartman** [google/or-tools#2586] [CP-SAT questions] [answered]
**Describe the problem you are trying to solve.**
Can OR-Tools solve a Multiple Choice Knapsack Problem?

Benefit Cost optimization where the options that I have can be exclusive. So I have multiple sets of options where I can either pick one or none of the options and must spend money to maximize benefit to the system.

**Describe the solution you'd like**
Any heuristic solution with reasonable accuracy and run time.

**Describe alternatives you've considered**
I've written a couple algorithms myself, but know that they can be improved upon. I've looked into some research papers regarding the topic as well, but feel overwhelmed with what is the best option or path to take.

**Additional context**
To reiterate the problem, I have a set of actions (A1, A2, A3) that I can perform on a set of assets (B1, B2, B3) where each of these have a benefit and cost to performing the action on the asset. I can only perform at most one action on each asset though. I want to maximize the benefit to the entire network while spending the entire budget available.

I do not know if this is currently possible with OR-Tools currently and am waiting to join the Google Group currently. From what I understand this is a Multiple Choice Knapsack Problem and I did not see documentation on solving this kind of problem with this tool.

**Comment by Mizux** [comment, marked as answer]
Should be doable using the CP-SAT solver IMHO.

e.g. https://github.com/google/or-tools/blob/stable/ortools/sat/samples/multiple_knapsack_sat.py

Basically I would:
- create some boolean variables J_X_Y which will be true if perform task A_X on asset B_Y.
- create some constraints like sum of all J_i_1 for all i must be one (aka one action per asset) etc...
- then create an objective as the sum of  J_X_Y * benefit_x_y
- create a constraint J_X_Y * cost_X_Y <= budget for all X,Y...

**Reply by ruiyang00** [reply]
Hello @Mizux ,

I have a similar problem that I'm trying to solve using OR-Tools. Let me describe my problem. I have n items in which each item has its own/private choices that it could choose from. Each choice has a weight and value associated with it. I'm trying to maximize the profit of items picked into one knapsack. In this case, how should I create constraints to this matter?

Thanks

**Reply by Mizux** [reply]
create one item per item choice and add new constraint that sum of all choice item is equal or less to one ?
aka just add ~one more nested loop

**Reply by ruiyang00** [reply]
Hello @Mizux ,
Thanks for the reply. Could you elaborate more, please? I'm still a bit confused when creating the constraints. Lets say that we have values={1, 3, 4, 5}, weights={10, 20, 30, 40}, and capacity={40}. I have a total of 2 items(4 choices this 4 choices can also be seen as items. But the constrain is that pick the best ones out from {1,3} and {4,5}). Item 1 is only allowed to choose from {1, 3} and item 2 from {4, 5}. How should I create such constraints to achieve the objective? 

Thanks.
